go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+# Go SDK Workflow Interceptor Updates and OpenTelemetry Tracing Support
+
+This proposal is for updates to Go workflow interceptors with the following features:
+
+* Support for and implementation of OpenTelemetry tracing
+* Support idiomatic Go interceptor design
+* Provide feature parity with Java SDK and other missing features (see https://github.com/temporalio/sdk-go/issues/529)
+* Add activity interceptors
+* Make inbound interceptors future-proof for argument addition
+
+**These updates are known to NOT be backwards-compatible**
+
+## Interceptor Proposed Updates
+
+### WorkflowInterceptor.InterceptWorkflow
+
+Current:
+
+```go
+InterceptWorkflow(info *WorkflowInfo, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+```
+
+Proposal:
+
+```go
+InterceptWorkflow(ctx Context, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+```
+
+Why: Callers can use the context to get workflow info.
+
+### WorkflowXCallsInterceptor
+
+Proposal: Change the name to `WorkflowXInterceptor`. Why: `Calls` just adds noise (more accepted in Java, not Go).
+
+### WorkflowInboundCallsInterceptor.ExecuteWorkflow
+
+Current:
+
+```go
+ExecuteWorkflow(ctx Context, workflowType string, args ...interface{}) []interface{}
+```
+
+Proposal:
+
+```go
+type WorkflowInput struct {
+  Args []interface{}
+}
+
+ExecuteWorkflow(ctx Context, input *WorkflowInput) (interface{}, error)
+```
+
+Why:
+
+* Workflow type can be derived from the context _or_ the info first given to `InterceptWorkflow`
+* Workflows can only return a value + error or just an error and this makes it more explicit what the error is
+
+### WorkflowInboundCallsInterceptor.ProcessSignal
+
+Current:
+
+```go
+ProcessSignal(ctx Context, signalName string, arg interface{}) error
+```
+
+Proposal:
+
+```go
+type SignalInput struct {
+  SignalName string
+  Arg interface{}
+  EventID string
+}
+
+HandleSignal(ctx Context, input *SignalInput) error
+```
+
+Also, this will be called to handle the signal with the last invocation in the chain passing the signal to the workflow.
+
+Why:
+
+* `HandleSignal` matches Java nomenclature and new behavior better
+* Currently sending is sent separate and async, so the current version doesn't wrap the handling. Instead this will be
+  part of the async call and documentation will make clear it needs to be safe for concurrent invocation.
```

Comment:
After reading this point a few times, I failed to understand what you're suggesting.
Are you suggesting we wrap the handling of the signal? Is that even possible with channels?

Reply:
Updated to be clearer